remill/Arch/X86/Semantics/X87.cpp

```diff
@@ -418,67 +630,130 @@ DEF_ISEL(FIDIVR_ST0_MEMmem16int) = FIDIVR<M16>;
 namespace {
 
 template <typename T>
-DEF_SEM(FST, T dst, RF80 src) {
+DEF_FPU_SEM(FST, T dst, RF80 src) {
+  SetFPUIpOp();
   typedef typename BaseType<T>::BT BT;
-  Write(dst, BT(Read(src)));
+  auto res = CheckedFloatUnaryOp(state, [=] (float64_t x) {
+    return static_cast<BT>(x);
+  }, Read(src));
+  Write(dst, res);
   return memory;
 }
 
 template <typename T>
-DEF_SEM(FSTP, T dst, RF80 src) {
-  memory = FST<T>(memory, state, dst, src);
+DEF_FPU_SEM(FSTmem, T dst, RF80 src) {
+  SetFPUDp(dst);
+  return FST(memory, state, dst, src, pc, fop);
+}
+
+template <typename T>
+DEF_FPU_SEM(FSTP, T dst, RF80 src) {
+  memory = FST<T>(memory, state, dst, src, pc, fop);
   (void) POP_X87_STACK();
   return memory;
 }
 
-DEF_SEM(FISTm16, M16W dst, RF80 src) {
-  Write(dst, Unsigned(Float64ToInt16(FRoundUsingMode64(Read(src)))));
+template <typename T>
+DEF_FPU_SEM(FSTPmem, T dst, RF80 src) {
+  SetFPUDp(dst);
+  return FSTP(memory, state, dst, src, pc, fop);
+}
+
+//printf("input=%lx rounded=%lx back=%lx conv_cast=%lx conv_adj=%lx ie=%d pe=%d\n",
+//         reinterpret_cast<uint64_t &>(input),
+//         reinterpret_cast<uint64_t &>(rounded),
+//         reinterpret_cast<uint64_t &>(back),
+//         static_cast<int64_t>(conv_cast), static_cast<int64_t>(conv_adj),
+//         state.sw.ie, state.sw.pe);
+
```

Comment:
Was this `printf` statement left here commented out for later use?

Reply:
Good catch.